### PR TITLE
`SubClassedDataFrame.groupby().mean()` etc. use method of `SubClassedDataFrame`

### LINKED ISSUE #51757: BUG: `DataFrameSubClass.groupby()` doesn't use methods of subclass
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import numpy as np
import pandas as pd

# define Series and DataFrame subclasses which override mean()

class UnitSeries(pd.Series):
    @property
    def _constructor(self):
        return UnitSeries

    @property
    def _constructor_expanddim(self):
        return UnitDataFrame

    def mean(self, *args, **kwargs):
        return 1


class UnitDataFrame(pd.DataFrame):
    @property
    def _constructor(self):
        return UnitDataFrame

    @property
    def _constructor_expanddim(self):
        return UnitSeries

    def mean(self, *args, **kwargs):
        return 1

# create example data
params = ['a', 'b']
data = np.random.rand(4, 2)
udf = UnitDataFrame(data, columns=params)
udf['group'] = np.ones(4, dtype=int)
udf.loc[2:, 'group'] = 2

# calculate mean with and without groupby
print(udf.mean()) # prints 1 :)
print(udf.groupby('group').mean()) # not a 1 to be seen :(
print(udf.groupby('group').get_group().mean()) # prints 1 :)
```

### Issue Description

Suppose I create subclasses of `Series` and `DataFrame`, which override a method such as `mean()`. `groupby().mean()` doesn't use this method, returning to the original behaviour. However, `.groupby().get_group().mean()` works correctly.

### Expected Behavior

`SubClassedDataFrame.groupby().mean()` should use the mean method of the subclass, so in my example 

```
print(udf.groupby('group').mean())
```
should print
```
       a  b
group
1      1  1
2      1  1
```
instead of using the usual mean:
```
              a         b
group
1      0.688265  0.324780
2      0.178812  0.663476
```

### Installed Versions

<details>

/Users/adam/programming/env_pandas2dev/lib/python3.11/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")

INSTALLED VERSIONS
------------------
commit           : 7f53afc69ceaabc47b1224e7c0fdd93dfb3faa7d
python           : 3.11.2.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 22.3.0
Version          : Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:37 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6000
machine          : arm64
processor        : arm
byteorder        : little
LC_ALL           : en_GB.UTF-8
LANG             : None
LOCALE           : en_GB.UTF-8

pandas           : 0+untagged.31727.g7f53afc
numpy            : 1.24.2
pytz             : 2022.7.1
dateutil         : 2.8.2
setuptools       : 65.6.3
pip              : 23.0.1
Cython           : None
pytest           : 7.2.1
hypothesis       : 6.68.2
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.7.0
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.10.1
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : None
qtpy             : None
pyqt5            : None

</details>

### PR DESCRIPTION
**Pull request by AdamOrmondroyd**
- [x] closes #51757
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

I appreciate that changing the behaviour for all subclasses is overkill, so pointers on how to better do this would be great

### CODE REVIEW

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -687,7 +689,7 @@ def value_counts(
             # in a backward compatible way
             # GH38672 relates to categorical dtype
             ser = self.apply(
-                Series.value_counts,
+                self._obj_1d_constructor.value_counts,
```
Comment (jbrockmendel) [inline pandas/core/groupby/generic.py:692]: this is trouble bc in general we can't assume that _constructor is a class

Comment (Ormorod) [reply pandas/core/groupby/generic.py:692]: What else could it be?

Comment (jbrockmendel) [reply pandas/core/groupby/generic.py:692]: Geopandas has a callable that can dispatch to different classes.  @jorisvandenbossche has argued against deprecating allowing this.

File: pandas/tests/groupby/test_groupby_subclass.py [hunk 2]
```diff
@@ -103,3 +103,89 @@ def test_groupby_resample_preserves_subclass(obj):
...
+    assert udf.mean() == 1
+    assert all(udf.groupby("group").mean() == 1)
+    assert all(udf.groupby("group").median() == 1)
+    assert all(udf.groupby("group").std() == 1)
+    assert all(udf.groupby("group").var() == 1)
+    assert all(udf.groupby("group").sem() == 1)
+    assert all(udf.groupby("group").prod() == 1)
+    assert all(udf.groupby("group").min() == 1)
+    assert all(udf.groupby("group").max() == 1)
+    # print(udf.groupby('group').beans()) # AttributeError
```
Comment (lukashergt) [inline pandas/tests/groupby/test_groupby_subclass.py]: * Shouldn't it be 2, 3, 4, etc. for median, std, var, etc.?
* What about methods that don't appear in `pandas/core/groupby/groupby.py`, e.g.`skew()` or `kurt()`? Might be good to include some of those in the test, too?

Comment (Ormorod) [reply pandas/tests/groupby/test_groupby_subclass.py]: > * Shouldn't it be 2, 3, 4, etc. for median, std, var, etc.?

I'd already fixed this! I hadn't realised as I should also have been using `np.all`

> * What about methods that don't appear in `pandas/core/groupby/groupby.py`, e.g.`skew()` or `kurt()`? Might be good to include some of those in the test, too?

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (Ormorod) [reply pandas/core/groupby/generic.py]: I've added a check whether `self._obj_1d_constructor` is a `Series`

**Comment by jorisvandenbossche** [comment]
> I appreciate that changing the behaviour for all subclasses is overkill

I think that this might indeed be an issue for subclasses that don't override those methods. If we want to keep a fix like this, maybe checking whether the method is the one from pandas or not might be better (something like `self.obj.mean is Series.mean` instead of `type(self.obj) is Series`).

Because with the current change, if you have a subclass that doesn't influence how aggregations are done, you would get a big slowdown because of the current PR I think?

File: pandas/core/groupby/groupby.py [hunk 3]
```diff
@@ -1837,14 +1837,24 @@ def mean(
         Name: B, dtype: float64
         """
 
+        if not (self.obj.mean is Series.mean or self.obj.mean is DataFrame.mean):
```
Comment (Ormorod) [inline pandas/core/groupby/groupby.py:1840]: Checking for method equality breaks lots of the tests in `test_groupby.py`, I am still working on why this is

Comment (Ormorod) [reply pandas/core/groupby/groupby.py:1840]: @jorisvandenbossche would you expect this comparison to work? I've never compared methods in this way

File: pandas/core/groupby/groupby.py [hunk 4]
```diff
@@ -1837,14 +1837,26 @@ def mean(
         Name: B, dtype: float64
         """
 
+        if not (
+            type(self.obj).mean is Series.mean or type(self.obj).mean is DataFrame.mean
+        ):
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:1842]: What about a Boolean flag as a class attribute on Series / DataFrame that subclasses can override if they want to use their methods? Defaults to False.

**Review by rhshadrach — COMMENTED** [review]
The current implementation is ignoring args (e.g. numeric_only). 

It seems to me this is opening a can of worms that we will now need to support; any feature of groupby would then need to work for subclasses. I'm thinking in particular some of code behind agg, transform, apply. 

Does `.groupby(...).agg(["sum", "mean"])` work with this?

File: pandas/core/groupby/generic.py [hunk 5]
```diff
@@ -1335,7 +1343,7 @@ def aggregate(self, func=None, *args, engine=None, engine_kwargs=None, **kwargs)
                 else:
                     # GH#32040, GH#35246
                     # e.g. test_groupby_as_index_select_column_sum_empty_df
-                    result = cast(DataFrame, result)
+                    result = cast(self._obj_1d_constructor, result)
```
Comment (jbrockmendel) [inline pandas/core/groupby/generic.py:1346]: _constructor, not _1d_constructor.  Also i think this is OK as is

**Comment by jbrockmendel** [comment]
Is the WeightedDataFrame thing referenced in #51757 the motivating case here?

What if instead of this we did did something like

```
class DataFrame:
    @property
    def _gb_cls(self):
        return DataFrameGroupBy
```

and then had DataFrame.groupby return an instance of `_gb_cls`.  Then subclasses could override that.

**Comment by Ormorod** [comment]
> Is the WeightedDataFrame thing referenced in #51757 the motivating case here?
> 

Yes, our classes add a set of "weights" to the index which are used to calculate statistics such as the mean.

> What if instead of this we did did something like
> 
> ```
> class DataFrame:
>     @property
>     def _gb_cls(self):
>         return DataFrameGroupBy
> ```
> 
> and then had DataFrame.groupby return an instance of `_gb_cls`. Then subclasses could override that.

That would function, though feel this would be unexpected extra work for developers, so should be mentioned in the extending pandas section of the docs.

I was very surprised that groupby() doesn't just use the methods of its underlying class. If `DataFrame.mean()` and `Series.mean()` are themselves optimised, what is gained by GroupBy reimplementing them?

File: pandas/core/groupby/groupby.py (hunk 4 above)
Comment (Ormorod) [reply pandas/core/groupby/groupby.py:1842]: > Does .groupby(...).agg(["sum", "mean"]) work with this?

No: 

```
>>> udf.groupby('group').agg(['sum', 'mean'])
              a                   b
            sum      mean       sum      mean
group
1      0.322530  0.161265  0.769053  0.384527
2      1.714631  0.857315  0.407676  0.203838
```
Using my example where the mean should always be 1.

Comment (Ormorod) [reply pandas/core/groupby/groupby.py:1842]: However transform() does work:
```
>>> udf.groupby('group').transform('mean')
   a  b
0  1  1
1  1  1
2  1  1
3  1  1
```

**Comment by jbrockmendel** [comment]
> If DataFrame.mean() and Series.mean() are themselves optimised, what is gained by GroupBy reimplementing them?

Performance.  The paths that iterate over groups generally make a sorted copy of the DataFrame to make that iteration faster, but the iteration itself is still costly.

**Comment by Ormorod** [comment]
> > If DataFrame.mean() and Series.mean() are themselves optimised, what is gained by GroupBy reimplementing them?
> 
> Performance. The paths that iterate over groups generally make a sorted copy of the DataFrame to make that iteration faster, but the iteration itself is still costly.

but, if these behaviours only change for subclasses, then performance is only lost for subclasses which override mean etc, in which case surely it is preferable that the intended results are returned rather than a fast one?

File: pandas/core/groupby/generic.py (hunk 5 above)
Comment (Ormorod) [reply pandas/core/groupby/generic.py]: I have also wrestled with whether results should stick to the subclass or just be DataFrames

**Comment by Ormorod** [comment]
> > but, if these behaviours only change for subclasses, then performance is only lost for subclasses which override mean etc, in which case surely it is preferable that the intended results are returned rather than a fast one?
> 
> The question was why we use a cython path instead of iterating over DataFrame.mean. I gave an explanation of the status quo, not an argument against the fix suggested here.

Apologies, I misunderstood the difference you were highlighting.

How do you think I should proceed?

File: pandas/core/groupby/groupby.py (hunk 4 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:1842]: Thanks for checking. In my opinion that isn't necessarily a blocker for this, but we are expanding the behavior here and introducing new bugs that should get fixed.

If we are to move forward with some form of implementation for this, I think the test coverage here should be expanded to include at least agg, apply, and transform (even if they are currently producing incorrect results).

File: pandas/core/groupby/generic.py (hunk 5 above)
Comment (rhshadrach) [reply pandas/core/groupby/generic.py]: What does mypy resolve `self._constructor` to? I'd be surprised if it wasn't `Any`. You can check by adding `reveal_type(self._obj_1d_constructor)` on it's own line and then run mypy.

File: pandas/core/groupby/generic.py [hunk 6]
```diff
@@ -681,14 +683,20 @@ def value_counts(
 
         index_names = self.grouper.names + [self.obj.name]
 
+        constructor_1d = (
+            self._obj_1d_constructor
+            if isinstance(self._obj_1d_constructor, Series)
```
Comment (jbrockmendel) [inline pandas/core/groupby/generic.py:688]: i think `issubclass` rather than `isinstance`?

File: pandas/core/groupby/generic.py [hunk 7]
```diff
@@ -1290,9 +1298,9 @@ def aggregate(self, func=None, *args, engine=None, engine_kwargs=None, **kwargs)
         elif relabeling:
             # this should be the only (non-raising) case with relabeling
             # used reordered index of columns
-            result = cast(DataFrame, result)
+            result = cast(self.obj._constructor, result)
```
Comment (jbrockmendel) [inline pandas/core/groupby/generic.py:1301]: these will be wrong if _constructor is not a class

**Comment by jbrockmendel** [comment]
> How do you think I should proceed?

Unless there is a viable way to handle your use case with an ExtensionArray (i.e. the weights would be somehow included with the values, not at the DataFrame level), then the approach here seems fine (albeit ugly)

File: doc/source/whatsnew/v2.1.0.rst [hunk 8]
```diff
@@ -223,6 +223,10 @@ Plotting
 - Bug in :meth:`Series.plot` when invoked with ``color=None`` (:issue:`51953`)
 -
 
+Groupby
+- Bug in :meth:`GroupBy.mean`, :meth:`GroupBy.median`, :meth:`GroupBy.std`, :meth:`GroupBy.var`, :meth:`GroupBy.sem`, :meth:`GroupBy.prod`, :meth:`GroupBy.min`, :meth:`GroupBy.max` don't use corresponding methods of subclasses of :class:`Series` or :class:`DataFrame` (:issue:`51757`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v2.1.0.rst:227]: i think this belongs below in the Groupby/resample/rolling section?

File: pandas/core/groupby/groupby.py [hunk 9]
```diff
@@ -1892,9 +1904,20 @@ def median(self, numeric_only: bool = False):
         Series or DataFrame
             Median of values within each group.
         """
+        if not (
+            type(self.obj).median is Series.median
+            or type(self.obj).median is DataFrame.median
+        ):
```
Comment (jbrockmendel) [inline pandas/core/groupby/groupby.py:1910]: if this pattern is going to show up a lot, does it merit a decorator?

Comment (AdamOrmondroyd) [reply pandas/core/groupby/groupby.py:2040]: I've had a first pass at making a decorator, not sure how to deal with `engine` and `engine_kwargs`

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**